designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,627 @@
+
+```js
+exports.config = {
+    files: ["*.js"],
+    extends: require("eslint-config-standard"),
+    rules: {
+        semi: ["error", "always"]
+    }
+};
+```
+
+This config extends `eslint-config-standard` by assigning it to `extends`. You can also use an array for `extends` to extend from multiple configs:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    extends: [
+        require("eslint-config-standard"),
+        require("@me/eslint-config")
+    ],
+    rules: {
+        semi: ["error", "always"]
+    }
+};
+```
+
+#### Extending From `eslint:recommended` and `eslint:all`
+
+Both `eslint:recommended` and `eslint:all` can be represented as strings in the `extends` key. For example:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    extends: [
+        "eslint:recommended",
+        require("eslint-config-standard"),
+        require("@me/eslint-config")
+    ],
+    rules: {
+        semi: ["error", "always"]
+    }
+};
+```
+
+This config first extends `eslint:recommended` and then continues on to extend other configs.
+
+#### Overriding Configuration Based on File Patterns
+
+Whereas `.eslintrc` had an `overrides` key that made a hierarchical structure, the `eslint.config.js` file does not have any such hierarchy. Instead, users can return an array of configs that should be used. For example, consider this `.eslintrc` config:
+
+```yaml
+plugins: ["react"]
+rules:
+   react/jsx-uses-react: error
+    semi: error
+
+overrides:
+    - files: "*.md"
+      plugins: ["markdown"],
+      processor: "markdown/markdown"
+```
+
+This can be written in `eslint.config.js` as an array of two configs:
+
+```js
+exports.config = [
+    {
+        files: "*.js",
+        ruledefs: {
+            react: require("eslint-plugin-react").rules,
+        },
+        rules: {
+            "react/jsx-uses-react": "error",
+            semi: "error"
+        }
+    },
+    {
+        files: "*.md",
+        processor: require("eslint-plugin-markdown").processors.markdown
+    }
+];
+```
+
+When ESLint uses this config, it will check each `files` pattern to determine which configs apply. Any config with a `files` pattern matching the file to lint will be extracted and used (if multiple configs match, then those configs are merged to determine the final config to use). In this way, returning an array acts exactly the same as the array in `overrides`.
+
+If a config in the config array does not contain `files` or `ignores`, then that config applies to all files. For example:
+
+```js
+exports.config = [
+    {
+        globals: {
+            Foo: true
+        }
+    },
+    {
+        files: "*.js",
+        ruledefs: {
+            react: require("eslint-plugin-react").rules,
+        },
+        rules: {
+            "react/jsx-uses-react": "error",
+            semi: "error"
+        }
+    },
+    {
+        files: "*.md",
+        processor: require("eslint-plugin-markdown").processors.markdown
+    }
+];
```

Comment:
I think the "config applies to all files by default" behavior is potentially confusing when combined with the removal of the `--ext` option.

Suppose I'm using the config given here, and I'm linting a `src/` folder containing `foo.js`, `README.md`, and `index.html`. If I'm understanding the file resolution section correctly, adding `globals: { Foo: true }` to the config array would mean that there is now a matching entry for `index.html` (since the new entry matches all files), so `index.html` would suddenly start getting linted (probably resulting in a parsing error).

Reply:
There is no config with a `files` pattern matching `*.html`, so `index.html` will not get linted. A config without `files` applies to all files that are passed in so long as there is a config with a matching `files` pattern. A config without a `files` pattern is never matched on its own. I'll add some text to clarify that.